src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,753 @@
+      ]
+
+    parser :: Parser PactCalcConfig
+    parser = PactCalcConfig
+      <$> O.strOption
+            (O.long "pact-database-dir"
+             <> O.short 'd'
+             <> O.metavar "PACT DBDIR"
+             <> O.help "Source Pact database, which you wish to verify"
+            )
+      <*> rocksParser
+      <*> cwvParser
+      <*> targetsParser
+
+data PactImportConfig = PactImportConfig
+  { sourcePactDir :: FilePath
+  , targetPactDir :: Maybe FilePath
+  , rocksDir :: FilePath
+  , chainwebVersion :: ChainwebVersion
+  }
+
+pactImportMain :: IO ()
+pactImportMain = do
+  cfg <- O.execParser opts
+
+  let cids = allChains cfg.chainwebVersion
+  checkPactDbsExist cfg.sourcePactDir cids
+
+  C.withDefaultLogger YAL.Info $ \logger -> do
+    -- Get the highest common blockheight across all chains.
+    (chainsContainingTarget, latestCommonHeight) <- resolveTarget logger cids cfg.sourcePactDir Latest
+
+    let (bh, expectedChainHashes) =
+          -- Find the first element of 'grands' such that
+          -- the blockheight therein is less than or equal to
+          -- the argument latest common blockheight.
+          case List.find (\g -> latestCommonHeight >= fst g) grands of
+            Nothing -> error "pact-import: no snapshot found"
+            Just (b, s) -> (b, Map.map (\es -> (es.pactHash, es.blockHeader)) s)
+
+    chainHashes <- do
+      let targets = TargetAll (Set.singleton bh)
+      computeGrandHashesAt logger (Set.toList chainsContainingTarget) cfg.sourcePactDir cfg.rocksDir targets cfg.chainwebVersion >>= \case
+        [(_, chainHashes)] -> do
+          pure chainHashes
+        [] -> do
+          error "pact-import: computeGrandHashesAt unexpectedly returned 0 blocks"
+        _ -> do
+          error "pact-import: computeGrandHashesAt unexpectedly returned multiple blocks"
+
+    let deltas = P.diff expectedChainHashes chainHashes
+
+    forM_ (Map.toAscList deltas) $ \(cid, delta) -> do
+      case delta of
+        P.Same _ -> pure ()
+        P.Old _ -> error "pact-import: internal logic error: chain mismatch"
+        P.New _ -> error "pact-import: internal logic error: chain mismatch"
+        P.Delta (eHash, eHeader) (hash, header) -> do
+          when (header /= eHeader) $ do
+            putStrLn $ unlines
+              [ "Chain " <> Text.unpack (chainIdToText cid)
+              , "Block Header mismatch"
+              , "  Expected: " <> show (_blockHash eHeader)
+              , "  Actual:   " <> show (_blockHash header)
+              ]
+
+          when (hash /= eHash) $ do
+            putStrLn $ unlines
+              [ "Chain " <> Text.unpack (chainIdToText cid)
+              , "Grand Hash mismatch"
+              , "  Expected: " <> Text.unpack (hex eHash)
+              , "  Actual:   " <> Text.unpack (hex hash)
+              ]
+    when (Map.size deltas > 0) exitFailure
+
+    logFunctionText logger Info "Hashes aligned"
+
+    -- TODO: drop things after the verified height?
```

Comment:
Imperative question. Should we import a tool that has history _after_ the latest height we can verify? Probably not, at least by default. Arguably there should be a flag, `--unsafe-keep-after-verified N` that takes a number of blocks N to allow after the verified height. maybe it shouldn't take an N at all, but just be a switch (boolean) flag.

Reply:
@edmundnoble @larskuhtz